File: src/main/java/com/refinedmods/refinedstorage/inventory/item/UpgradeItemHandler.java
package com.refinedmods.refinedstorage.inventory.item;

import com.refinedmods.refinedstorage.inventory.item.validator.UpgradeItemValidator;
import com.refinedmods.refinedstorage.item.UpgradeItem;
import net.minecraft.world.item.ItemStack;

public class UpgradeItemHandler extends BaseItemHandler {
    public UpgradeItemHandler(int size, UpgradeItem.Type... supportedUpgrades) {
        super(size);

        for (UpgradeItem.Type supportedUpgrade : supportedUpgrades) {
            addValidator(new UpgradeItemValidator(supportedUpgrade));
        }
    }

    public int getSpeed() {
        return getSpeed(9, 2);
    }

    public int getSpeed(int speed, int speedIncrease) {
        for (int i = 0; i < getSlots(); ++i) {
            ItemStack slot = getStackInSlot(i);

            if (slot.getItem() instanceof UpgradeItem && ((UpgradeItem) slot.getItem()).getType() == UpgradeItem.Type.SPEED) {
                speed -= speedIncrease;
            }
        }

        return speed;
    }

    public boolean hasUpgrade(UpgradeItem.Type type) {
        for (int i = 0; i < getSlots(); ++i) {
            ItemStack slot = getStackInSlot(i);

            if (slot.getItem() instanceof UpgradeItem && ((UpgradeItem) slot.getItem()).getType() == type) {
                return true;
            }
        }

        return false;
    }

    public int getUpgradeCount(UpgradeItem.Type type) {
        int upgrades = 0;

        for (int i = 0; i < getSlots(); ++i) {
            ItemStack slot = getStackInSlot(i);

            if (slot.getItem() instanceof UpgradeItem && ((UpgradeItem) slot.getItem()).getType() == type) {
                upgrades++;
            }
        }

        return upgrades;
    }

    public int getEnergyUsage() {
        int usage = 0;

        for (int i = 0; i < getSlots(); ++i) {
            ItemStack slot = getStackInSlot(i);

            if (slot.getItem() instanceof UpgradeItem) {
                usage += ((UpgradeItem) slot.getItem()).getType().getEnergyUsage();
            }
        }

        return usage;
    }

    public int getStackInteractCount() {
        return hasUpgrade(UpgradeItem.Type.STACK) ? 64 : 1;
    }

    @Override
    public int getSlotLimit(int slot) {
        return 1;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/item/UpgradeItemHandler.kt
package com.refinedmods.refinedstorage.inventory.item

import com.refinedmods.refinedstorage.inventory.item.validator.UpgradeItemValidator
import com.refinedmods.refinedstorage.item.UpgradeItem
import net.minecraft.item.ItemStack

open class UpgradeItemHandler(size: Int, vararg supportedUpgrades: UpgradeItem.Type) : BaseItemHandler(size) {
    val speed: Int
        get() = getSpeed(9, 2)

    fun getSpeed(speed: Int, speedIncrease: Int): Int {
        var speed = speed
        for (i in 0 until getSlots()) {
            val slot: ItemStack = getStackInSlot(i)
            if (slot.item is UpgradeItem && (slot.item as UpgradeItem).type == UpgradeItem.Type.SPEED) {
                speed -= speedIncrease
            }
        }
        return speed
    }

    fun hasUpgrade(type: UpgradeItem.Type): Boolean {
        for (i in 0 until getSlots()) {
            val slot: ItemStack = getStackInSlot(i)
            if (slot.item is UpgradeItem && (slot.item as UpgradeItem).type == type) {
                return true
            }
        }
        return false
    }

    fun getUpgradeCount(type: UpgradeItem.Type): Int {
        var upgrades = 0
        for (i in 0 until getSlots()) {
            val slot: ItemStack = getStackInSlot(i)
            if (slot.item is UpgradeItem && (slot.item as UpgradeItem).type == type) {
                upgrades++
            }
        }
        return upgrades
    }

    val energyUsage: Int
        get() {
            var usage = 0
            for (i in 0 until getSlots()) {
                val slot: ItemStack = getStackInSlot(i)
                if (slot.item is UpgradeItem) {
                    usage += (slot.item as UpgradeItem).type.energyUsage
                }
            }
            return usage
        }
    open val stackInteractCount: Int
        get() = if (hasUpgrade(UpgradeItem.Type.STACK)) 64 else 1

    fun getSlotLimit(slot: Int): Int {
        return 1
    }

    init {
        for (supportedUpgrade in supportedUpgrades) {
            addValidator(UpgradeItemValidator(supportedUpgrade))
        }
    }
}